### PR TITLE
Add MyPy typing to accelerators

### PR DESCRIPTION
**Pull request by justusschock**
## What does this PR do?

Fixes #<issue_number>

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/accelerators/accelerator.py [hunk 1]
```diff
@@ -24,6 +24,11 @@
 from pytorch_lightning.utilities.distributed import all_gather_ddp_if_available
 from pytorch_lightning.utilities.enums import AMPType, LightningEnum
 
+if TYPE_CHECKING:
+    from torch.cuda.amp import GradScaler
```
Comment (Borda) [inline pytorch_lightning/accelerators/accelerator.py:28]: this can be imported normally, right?

File: pytorch_lightning/accelerators/accelerator.py [hunk 2]
```diff
@@ -24,6 +24,11 @@
...
+
+    from pytorch_lightning.trainer.trainer import Trainer
```
Comment (Borda) [inline pytorch_lightning/accelerators/accelerator.py:30]: do you need this even you annotate it with a string?

File: pytorch_lightning/accelerators/accelerator.py [hunk 3]
```diff
@@ -213,23 +221,29 @@ def predict(self, args):
...
         """A hook to do something at the end of the test step
 
         Args:
             output: the output of the test step
         """
         return self.training_type_plugin.test_step_end(output)
 
-    def validation_step_end(self, output):
+    def validation_step_end(
+        self, output: Union[torch.Tensor, Dict[str, torch.Tensor], None]
```
Comment (Borda) [inline pytorch_lightning/accelerators/accelerator.py:245]: I think that the recommendation for `Union[..., None]` or `Optional[Union[...]]`

File: pytorch_lightning/accelerators/accelerator.py [hunk 4]
```diff
@@ -351,26 +367,22 @@ def precision(self) -> int:
         return self.precision_plugin.precision
 
     @property
-    def scaler(self):
-        if hasattr(self.precision_plugin, "scaler"):
-            return self.precision_plugin.scaler
+    def scaler(self) -> Optional['GradScaler']:
 
-        return None
+        return getattr(self.precision_plugin, 'scaler', None)
```
Comment (Borda) [inline pytorch_lightning/accelerators/accelerator.py:372]: was this a bug?

File: pytorch_lightning/accelerators/accelerator.py (hunk 2 above)
Comment (justusschock) [reply pytorch_lightning/accelerators/accelerator.py:30]: Yes, because if we annotate with the plain class it still has to be imported globally, since else there are undefined variables during Runtime. MyPy then reads those strings as class variables

File: pytorch_lightning/accelerators/accelerator.py (hunk 3 above)
Comment (justusschock) [reply pytorch_lightning/accelerators/accelerator.py:245]: I don't get you

File: pytorch_lightning/accelerators/accelerator.py (hunk 4 above)
Comment (justusschock) [reply pytorch_lightning/accelerators/accelerator.py:372]: No, this behaves the same, but It get's around MyPys error "Precision Plugin has no attribute scaler", which we covered by returning None. But Mypy was not able to get this.

File: pytorch_lightning/accelerators/accelerator.py (hunk 1 above)
Comment (justusschock) [reply pytorch_lightning/accelerators/accelerator.py:28]: It could, but it is not needed during runtime, so why import it then?

File: pytorch_lightning/accelerators/accelerator.py (hunk 3 above)
Comment (justusschock) [reply pytorch_lightning/accelerators/accelerator.py:245]: You think I should replace this? Optional is just an alias and is convenient when having only one other type. I think the current way is fine

Comment (Borda) [reply pytorch_lightning/accelerators/accelerator.py:245]: I am fine with it, just a side comment

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/accelerators/accelerator.py [hunk 5]
```diff
@@ -351,26 +367,22 @@ def precision(self) -> int:
...
         Returns state of an optimizer. Allows for syncing/collating optimizer state from processes in custom
         plugins.
         """
-        if self.training_type_plugin and hasattr(self.training_type_plugin, "optimizer_state"):
-            return self.training_type_plugin.optimizer_state(optimizer)
-        return optimizer.state_dict()
+        return getattr(self.training_type_plugin, 'optimizer_state', lambda x: x.state_dict())(optimizer)
 
-    def on_save(self, checkpoint):
+    def on_save(self, checkpoint: Dict[str, Union[Any, torch.Tensor]]) -> Dict[str, Union[Any, torch.Tensor]]:
```
Comment (carmocca) [inline pytorch_lightning/accelerators/accelerator.py:385]: Same comment about `Union[Any, torch.Tensor]`

File: pytorch_lightning/accelerators/accelerator.py [hunk 6]
```diff
@@ -134,11 +139,14 @@ def batch_to_device(self, batch: Any, device: Optional[torch.device] = None) ->
...
 
-    def on_train_start(self):
+    def on_train_start(self) -> None:
         """Hook to do something upon the training start"""
         pass
 
-    def training_step(self, args):
+    def training_step(
+        self,
+        args: List[Union[Any, int]],
```
[the remark this answers is not in the document]
Comment (justusschock) [reply pytorch_lightning/accelerators/accelerator.py:148]: Yes, it is equal to just Any, but when you have some additional info about what the type could potentially be, IDEs use that for autocomplete.

File: pytorch_lightning/accelerators/accelerator.py (hunk 5 above)
Comment (justusschock) [reply pytorch_lightning/accelerators/accelerator.py:385]: Same answer as above :)

File: pytorch_lightning/accelerators/accelerator.py [hunk 7]
```diff
@@ -385,7 +397,9 @@ def broadcast(self, obj: object, src: int = 0) -> object:
         """
         return self.training_type_plugin.broadcast(obj, src)
 
-    def all_gather(self, tensor: Union[torch.Tensor], group: Optional[Any] = None, sync_grads: bool = False):
+    def all_gather(
+        self, tensor: Union[torch.Tensor], group: Optional[Any] = None, sync_grads: bool = False
```
Comment (akihironitta) [inline pytorch_lightning/accelerators/accelerator.py:401]: Could this be simplified?:
```suggestion
        self, tensor: torch.Tensor, group: Any = None, sync_grads: bool = False
```

File: pytorch_lightning/accelerators/tpu.py [hunk 8]
```diff
@@ -27,10 +31,14 @@ def setup(self, trainer, model):
...
 
-    def run_optimizer_step(self, optimizer: Optimizer, optimizer_idx: int, lambda_closure: Callable, **kwargs):
+    def run_optimizer_step(
+        self, optimizer: Optimizer, optimizer_idx: int, lambda_closure: Callable, **kwargs: Any
+    ) -> None:
         xm.optimizer_step(optimizer, optimizer_args={'closure': lambda_closure, **kwargs})
 
-    def all_gather(self, tensor: Union[torch.Tensor], group: Optional[Any] = None, sync_grads: bool = False):
+    def all_gather(
+        self, tensor: Union[torch.Tensor], group: Optional[Any] = None, sync_grads: bool = False
```
Comment (akihironitta) [inline pytorch_lightning/accelerators/tpu.py:40]: Same as above
```suggestion
        self, tensor: torch.Tensor, group: Any = None, sync_grads: bool = False
```

File: pytorch_lightning/accelerators/accelerator.py (hunk 7 above)
Comment (justusschock) [reply pytorch_lightning/accelerators/accelerator.py:401]: For the first part yes, but I'd still have the Optional here, since this IMO helps with readibility, since it means it could be an object but also None

**Review by awaelchli — APPROVED** [review]
very nice :)

**Review by carmocca — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/accelerators/tpu.py [hunk 9]
```diff
@@ -1,4 +1,4 @@
-from typing import Any, Callable, Optional, Union
+from typing import Any, Callable, Optional, TYPE_CHECKING, Union
```
Comment (rohitgr7) [inline pytorch_lightning/accelerators/tpu.py:1]:
```suggestion
from typing import Any, Callable, Optional, TYPE_CHECKING
```

File: pytorch_lightning/accelerators/accelerator.py [hunk 10]
```diff
@@ -11,7 +11,7 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-from typing import Any, Callable, Iterable, Optional, Union
+from typing import Any, Callable, Dict, Iterable, List, Optional, Sequence, Tuple, TYPE_CHECKING, Union
```
Comment (rohitgr7) [inline pytorch_lightning/accelerators/accelerator.py:14]:
```suggestion
from typing import Any, Callable, Dict, Iterable, List, Optional, Sequence, TYPE_CHECKING, Union
```

**Review by rohitgr7 — APPROVED** [review]
